### PR TITLE
[Kunlun] add collective ops for multi XPU cards training

### PR DESCRIPTION
**Pull request by vslyu**
### PR types

New features
### PR changes

OPs
### Describe

- allreduce_sum/prod/max/min op for xpu
- reduce_sum/prod/max/min op for xpu

example:

fleetrun --xpus 0,1 example.py

```
import numpy as np
import paddle
from paddle.distributed import init_parallel_env

paddle.set_device('xpu:%d'%paddle.distributed.ParallelEnv().dev_id)
init_parallel_env()
if paddle.distributed.ParallelEnv().local_rank == 0:
    np_data = np.array([[4, 5, 6], [4, 5, 6]])
else:
    np_data = np.array([[1, 2, 3], [1, 2, 3]])
data = paddle.to_tensor(np_data)
paddle.distributed.reduce(data, 0)
out = data.numpy()
# [[5, 7, 9], [5, 7, 9]]
```

### CODE REVIEW

File: paddle/fluid/operators/collective/c_allreduce_op.h [hunk 1]
```diff
@@ -105,6 +113,71 @@ class CAllReduceOpCPUKernel : public framework::OpKernel<T> {
...
+  void Compute(const framework::ExecutionContext& ctx) const override {
+#if defined(PADDLE_WITH_XPU_BKCL)
+    auto in = ctx.Input<framework::Tensor>("X");
+    auto out = ctx.Output<framework::Tensor>("Out");
+
+    auto place = ctx.GetPlace();
+    BKCLDataType dtype = platform::ToBKCLDataType(in->type());
+    int64_t numel = in->numel();
+    const void* sendbuff = in->data<void>();
+    out->Resize(in->dims());
```
Comment (QingshuChen) [inline paddle/fluid/operators/collective/c_allreduce_op.h:128]: resize需要*sizeof(datatype)吗？

File: paddle/fluid/operators/collective/c_allreduce_op.h [hunk 2]
```diff
@@ -105,6 +113,71 @@ class CAllReduceOpCPUKernel : public framework::OpKernel<T> {
...
+      case kRedProd:
+        bkcl_red_type = BKCL_PRODUCT;
+        break;
+
+      default:
+        PADDLE_THROW(platform::errors::InvalidArgument(
+            "Invalid reduce type: %d", red_type));
+    }
+
+    // PADDLE_ENFORCE_CUDA_SUCCESS(platform::dynload::ncclAllReduce(
```
Comment (QingshuChen) [inline paddle/fluid/operators/collective/c_allreduce_op.h]: 无用的注释

File: paddle/fluid/operators/collective/c_allreduce_op.h (hunk 1 above)
Comment (vslyu) [reply paddle/fluid/operators/collective/c_allreduce_op.h:128]: 不需要，tensor自己封装的函数

File: paddle/fluid/operators/collective/c_allreduce_max_op_xpu.cc [hunk 3]
```diff
@@ -0,0 +1,28 @@
+/* Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (QingshuChen) [inline paddle/fluid/operators/collective/c_allreduce_max_op_xpu.cc:1]: 加ifdef PADDLE_WITH_XPU

File: paddle/fluid/operators/collective/c_allreduce_min_op_xpu.cc [hunk 4]
```diff
@@ -0,0 +1,28 @@
+/* Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+
```
Comment (QingshuChen) [inline paddle/fluid/operators/collective/c_allreduce_min_op_xpu.cc:2]: 同上

**Pull request closed without merging**